.github/workflows/benchmarks.yml

```diff
@@ -5,8 +5,30 @@ permissions: {}
 jobs:
   telio-firewall:
       runs-on: ubuntu-22.04
+      env:
+        CARGO_TARGET_DIR: "/home/runner/work/libtelio/libtelio/target/criterion"
       steps:
         - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
+        - name: Install critcmp for benchmark comparison
+          run: cargo install critcmp
         - name: Telio firewall benchmarks
           working-directory: crates/telio-firewall
-          run: cargo bench --features test_utils --bench firewall_bench "64" -- --warm-up-time 1 --measurement-time 1
+          run: cargo bench --features test_utils --bench firewall_bench "64" -- --warm-up-time 1 --measurement-time 1 --save-baseline current
+        - name: Benchmark baseline
+          working-directory: crates/telio-firewall
+          run: | # v4.0.0 is just a placeholder
+            git fetch --all
+            git checkout v5.0.0-rc3
+            cargo bench --features test_utils --bench firewall_bench "64" -- --warm-up-time 1 --measurement-time 1 --save-baseline baseline
+        - name: Compare benchmarks
+          working-directory: crates/telio-firewall
+          run: |
+            result=$(critcmp current baseline -t 5)
+            output=$($output | awk '{if (NR > 2) { if ($10 == 1.00 ||  $14 == 1.00) { if ($10 < $14) {exit 1} } else { if ($13 < $17) {exit 1}}}}')
+            if [ -z "$output" ]; then
+              echo "No performance regression detected."
+            else
+              echo "Performance regression detected! Failing the build."
```

Comment:
This is run on the public github owned runner - we have no control over how busy is the machine any any point in time so I'm not sure that those numbers are meaningful.

Reply:
Small differences might not come up, but a slowdown should still be detected, something which would either way fall through as we don't regress.
Increasing the number of runs and taking average, or well self-hosting would be alternatives.

Reply:
**Maybe**, but it also means that this is prone to being flaky. Imo, without runner physical machine that is not being used for anything when the benchmarks are being run, this shouldn't be merged in.

Reply:
We should then think about how to do that, as the same problem would come in for any other performance tests even with boringtun.